### ISSUE TITLE
New Rule: discourage type assertions in favor of type guards

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#1228]
Name: `no-type-assertion`
- lines up with existing `no-angle-bracket-type-assertion` rule

Failing code:

``` ts
function foo(myArg: any) {
  if (myArg.bar != null) {
    doTheThing(myArg as IBar);
  } else if (myArg.baz != null) {
    anotherThing(myArg as IBaz);
  } else {
    ...
  }
}
```

Passing code:

``` ts
function foo(myArg: any) {
  if (isBar(myArg)) {
    doTheThing(myArg);
  } else if (isBaz(myArg)) {
    anotherThing(myArg);
  } else {
    ...
  }
}

function isBar(x: any): x is IBar {
  ...
}

function isBaz(x: any): x is IBaz {
  ...
}
```

**Comment by jskrzypek** [comment]
+1

**Comment by JoshuaKGoldberg** [comment]
Relevant: discussion in #2040 on bringing in the equivalent rule from tslint-consistent-codestyle.

IMO it'd be better to name this as `type-assertion` and allow configuring as preferring `as` (default) or preferring <>.

Note that as ajafff mentioned, copied code without substantial changes needs to retain the original license.

**Comment by JoshuaKGoldberg** [comment]
Relevant: discussion in #2040 on bringing in the equivalent rule from tslint-consistent-codestyle.

IMO it'd be better to name this `type-assertion` and allow configuring to:
* Ban type guards (default)
* Allow `as Type`
* Allow `<Type>

Note that as ajafff mentioned, copied code without substantial changes needs to retain the original license.

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
